search: support selecting symbol kind

rvantonder opened the pull request:
Allows to specify `select:symbol.<kind>`. Part of https://github.com/sourcegraph/sourcegraph/issues/18002.

https://user-images.githubusercontent.com/888624/109730916-8bb2af00-7b77-11eb-9cd9-a52386b63e91.mp4

I've chosen `kind` to be based on the reduced set of LSP-like values, and not the varied and aliased ctags values that we actually label symbols as internally. E.g., there's little benefit of supporting both `func` and `function` and we can cut down on suggested values when I support this in frontend.

I will add another unit test and integration test later but this is ready for review.

rvantonder commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -201,9 +201,15 @@ func (fm *FileMatchResolver) Select(t filter.SelectPath) SearchResultResolver {
-			fm.FileMatch.LineMatches = nil
+			fm.FileMatch.LineMatches = nil // Only return symbol match if symbols exist
+			if len(t.Fields) > 0 {
+				filteredSymbols := SelectKind(fm.FileMatch.Symbols, t.Fields[0])
```
Only one level of field nesting supported/validated. I declare this is adequate for now :-)

rvantonder commented on cmd/frontend/graphqlbackend/select_symbols.go:
```diff
@@ -0,0 +1,93 @@
+package graphqlbackend
```
This file can be moved outside `graphqlbackend` after https://github.com/sourcegraph/sourcegraph/pull/18794 I think, but not possible right now.

camdencheek commented on internal/search/filter/select.go:
```diff
@@ -29,9 +33,52 @@ var validSelectors = map[SelectType]struct{}{
 	Symbol:     {},
 }
 
+var validFields = map[SelectType]interface{}{
```
I think the top level type can be more strictly defined as `map[SelectType]map[string]interface{}`, which will let us get rid of the cast on line 79

camdencheek commented on internal/search/filter/select.go:
```diff
@@ -29,9 +33,52 @@ var validSelectors = map[SelectType]struct{}{
 	}
-	return SelectPath{SelectType(s)}, nil
+	if len(fields) > 0 {
+		if _, ok := validFields[SelectType(selector)].(map[string]interface{})[fields[0]]; !ok {
```
thoughts on combining `validFields` and `validSelectors`?

camdencheek reviewed (approved):
Awesome! Looks great

rvantonder commented:
Moved the select symbols logic to `internal/search` and added unit test for selecting. There's more to simplify here and I'm focused on symbols and types for the rest of the day, so probably best to  avoid touching that stuff until I'm done :-)
